packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -473,6 +473,53 @@ export default class UserStorageController extends BaseController<
     );
   }
 
+  #isLocked = false;
+
+  readonly #pendingOperations: (() => Promise<unknown> | unknown)[] = [];
+
+  /**
+   * Locks the controller, preventing operations that are subsequently called with `executeWithLockHandling` from executing.
+   * These operations will be queued until the controller is unlocked.
+   */
+  lock() {
+    this.#isLocked = true;
+  }
+
+  /**
+   * Unlocks the controller and executes all pending operations that were queued using the `executeWithLockHandling` method while the controller was locked.
+   * This method will also clear the pending operations queue.
+   * If any of the operations fail, the error will be propagated to the caller. The following operations will not be executed and the queue will be cleared.
+   */
+  async unlock() {
+    try {
+      for (const operation of this.#pendingOperations) {
+        await operation();
+      }
+    } finally {
+      this.#pendingOperations.length = 0;
+      this.#isLocked = false;
+    }
+  }
+
+  /**
+   * Executes or enqueues an operation based on the lock state of the controller.
+   * If the controller is locked, the operation will be added to the pending operations queue.
+   * If the controller is not locked, the operation will be executed immediately.
+   *
+   * @param operation - The operation to execute. This should be a function that returns a promise.
+   * @returns The result of the operation.
+   */
+  async executeWithLockHandling<T>(
```

Comment:
This method should only return void promises.
Since operations are potentially queued, it's unreasonable to promise a non `void` response. This is already reflected in the `T | void` response type, but it's non deterministic for the caller, and not a true error scenario either.